core/server/data/migrations/versions/3.29/02-add-member-constraints.js

```diff
@@ -0,0 +1,103 @@
+const logging = require('../../../../../shared/logging');
+
+module.exports = {
+    config: {
+        transaction: true
+    },
+
+    async up({transacting: knex}) {
+        if (knex.client.config.client !== 'mysql') {
+            return logging.warn('Skipping member tables index creation - database is not MySQL');
+        }
+
+        const [membersLabelsIndexes] = await knex.raw('SHOW INDEXES FROM members_labels');
+        const [membersStripeCustomersIndexes] = await knex.raw('SHOW INDEXES FROM members_stripe_customers');
+        const [membersStripeCustomersSubscriptionsIndexes] = await knex.raw('SHOW INDEXES from members_stripe_customers_subscriptions');
+
+        if (membersLabelsIndexes.find(index => index.Key_name === 'members_labels_member_id_foreign')) {
```

Comment:
I think we gonna need slightly more complicated logic for this :( There should never be a case of the index not existing - but I think we need to make sure that we add the `ON DELETE CASCADE` - not sure how easy it is to check for that? Though, maybe we should be able to "know" the current state and just unconditionally add it?

Reply:
> not sure how easy it is to check for that?

It's possible to query for it. You can either do a query for all constraints on the db and filter, or query for the specific one you're interested in:

```sql
SELECT * FROM information_schema.REFERENTIAL_CONSTRAINTS WHERE CONSTRAINT_SCHEMA="${dbName}";
```

```sql
SELECT * FROM information_schema.REFERENTIAL_CONSTRAINTS WHERE CONSTRAINT_SCHEMA="${dbName}" AND CONSTRAINT_NAME="members_labels_member_id_foreign";
```

```
mysql> SELECT * FROM information_schema.REFERENTIAL_CONSTRAINTS WHERE CONSTRAINT_SCHEMA="ghost_dev" AND CONSTRAINT_NAME="members_labels_member_id_foreign" \G;
*************************** 1. row ***************************
       CONSTRAINT_CATALOG: def
        CONSTRAINT_SCHEMA: ghost_dev
          CONSTRAINT_NAME: members_labels_member_id_foreign
UNIQUE_CONSTRAINT_CATALOG: def
 UNIQUE_CONSTRAINT_SCHEMA: ghost_dev
   UNIQUE_CONSTRAINT_NAME: PRIMARY
             MATCH_OPTION: NONE
              UPDATE_RULE: RESTRICT
              DELETE_RULE: CASCADE
               TABLE_NAME: members_labels
    REFERENCED_TABLE_NAME: members
1 row in set (0.00 sec)
```

I can update the migration to query that and adjust to add/remove the delete cascade.

Reply:
Ah cool okay :relaxed: 

I _think_ that makes sense? 

I updated the `members_labels` table in the schema to have the delete cascade here: https://github.com/TryGhost/Ghost/pull/12105/commits/44ff90c50b3c5d954eda819e091d53ee143a6df4, I added it for _both_ member_id and label_id - does that make sense to you?

If not, I'll remove it, but if so, I think this migration will need to handle both indexes on the `members_labels` table :grimacing:

Reply:
> I added it for both member_id and label_id - does that make sense to you?

My only concern is whether our bookshelf plugins run into any issues when deleting labels. I'm guessing not because it deletes all the `members_labels` rows before deleting the label itself but it would be good to confirm.

Otherwise, makes sense, yeah 😄

Reply:
Updated add/remove ON DELETE CASCADE rather than dropping the whole constraint for members_labels constraints.